packages/deploy/src/index.ts

```diff
@@ -103,6 +103,55 @@ export async function getSpec(path: string) {
 
 // =============================================
 
+async function getAllSpecJobs(
+  config: DeployConfig,
+  logger: Logger
+): Promise<SpecJob[]> {
+  const jobs: SpecJob[] = [];
+
+  try {
+    const [state, spec] = await Promise.all([
+      getState(config.statePath),
+      getSpec(config.specPath),
+    ]);
+
+    for (const [workflowKey, workflow] of Object.entries(spec.doc.workflows)) {
+      if (workflow.jobs) {
+        for (const [jobKey, specJob] of Object.entries(workflow.jobs)) {
+          const stateJob = state.workflows[workflowKey]?.jobs[jobKey];
+          stateJob &&
+            jobs.push({
+              id: stateJob.id,
+              name: specJob.name,
+              adaptor: specJob.adaptor,
+              body: specJob.body,
+            });
+        }
+      }
+    }
+  } catch (error: any) {
+    logger.debug(`Could not read the spec and state: ${error.message}`);
+  }
+
+  return jobs;
+}
+
+export async function updatePulledSpecJobBodyPath(
```

Comment:
Can I ask for a better function name here? And maybe on `addSpecJobBodyPath` ?

This is really complex stuff and it's taken me ages to get any kind of feeling for what's happening here :sweat_smile:  

Basically what these functions are doing is pulling out job bodies from the newly pulled spec and writing them to disk if the state (?) includes a path. Right?

So something like `extractJobsToDisk`  might be a better name for `addSpecJobBodyPath`

Probably `updatePulledSpecJobBodyPath` should be moved to a new file `packages/deploy/src/pull.ts`, since it only affects pulls the context helps understandability. A name like `syncRemoteSpec` or something might be a good general name for what this is trying to do